### PR TITLE
Remove C-style arrays from smart pointers exercise.

### LINKED ISSUE #163: Use std::vector in smartPointers and stl exercises ?
It is my strong feeling that the use of raw C arrays makes the syntax uselessly complex, and divert the learners from what we want to demonstrate. It would rather use std::vector, provided it has been previously presented in the morning lessons.

### PR DESCRIPTION
**Pull request by hageboeck**
Fix #163.

### CODE REVIEW

File: code/smartPointers/smartPointers.cpp [hunk 1]
```diff
@@ -30,21 +31,21 @@
  */
 
 // Declare a function to do something with the data. No need to change it.
-double sumEntries(const double* data, std::size_t size) {
-    if (size > 200)
-        throw std::invalid_argument("I only want to sum 200 numbers or less.");
+template<typename T>
+double sumEntries(const T * data) {
```
Comment (bernhardmgruber) [inline code/smartPointers/smartPointers.cpp:35]: I was sure there is an error when you call `data->begin()` later until I realized you pass a `std::array<T>*`. Consider renaming the template paramter to make this clearer:
```suggestion
template<typename Range>
double sumEntries(const Range * data) {
```

Comment (hageboeck) [reply code/smartPointers/smartPointers.cpp:35]: > I was sure there is an error when you call data->begin()

I compile before I commit. 🙂
But it's a good idea to call it range, so we'll go for that!

File: code/smartPointers/smartPointers.cpp [hunk 2]
```diff
@@ -30,21 +31,21 @@
  */
 
 // Declare a function to do something with the data. No need to change it.
-double sumEntries(const double* data, std::size_t size) {
-    if (size > 200)
-        throw std::invalid_argument("I only want to sum 200 numbers or less.");
+template<typename Range>
+double sumEntries(const Range * data) {
```
Comment (sponce) [inline code/smartPointers/smartPointers.cpp:35]: Maybe let's be slightly simpler by avoiding the template and using std::span ?

**Review by sponce — APPROVED** [review]

**Review by chavid — APPROVED** [review]
Ok for me.

**Comment by sponce** [comment]
somehow checks are failing because we try to compile for C++11 something that clearly needs C++20. Not sure what is wrong

File: code/smartPointers/Makefile [hunk 3]
```diff
@@ -5,7 +5,7 @@ clean:
 	rm -f *o *so smartPointers *~ smartPointers.sol
 
 % : %.cpp
-	$(CXX) -g -std=c++17 -Wall -Wextra -o $@ $<
+	$(CXX) -g -std=c++20 -Wall -Wextra -o $@ $<
 
 %.sol : solution/%.sol.cpp
-	$(CXX) -g -std=c++17 -Wall -Wextra -o $@ $<
+	$(CXX) -g -std=c++20 -Wall -Wextra -o $@ $<
```
Comment (bernhardmgruber) [inline code/smartPointers/Makefile:11]: You may want to apply the same changes to `CMakeLists.txt` as well.

Comment (hageboeck) [reply code/smartPointers/Makefile:11]: Done, all green. Merging now.

**Pull request merged**